index/index.go

```diff
@@ -0,0 +1,393 @@
+var ErrNotAuthorized = errors.New("user not authorized")
+
+type SecureIndex struct {
+	keyProvider key.Provider
+	ioProvider  io.Provider
+	idProvider  id.Provider
+
+	indexKey []byte
+}
+
+// NewSecureIndex creates a SecureIndex which is used to manage keyword/identifier pairs.
+func NewSecureIndex(keyProvider key.Provider, ioProvider io.Provider, idProvider id.Provider) (SecureIndex, error) {
+	keys, err := keyProvider.GetKeys()
+	if err != nil {
+		return SecureIndex{}, err
+	}
+
+	if len(keys.IEK) != MasterKeyLength {
+		return SecureIndex{}, ErrInvalidMasterKeyLength
+	}
+
+	return SecureIndex{
+		keyProvider: keyProvider,
+		ioProvider:  ioProvider,
+		idProvider:  idProvider,
+		indexKey:    keys.IEK,
+	}, nil
+}
+
+// Add adds a keyword/identifier pair to the secure index.
+func (i *SecureIndex) Add(token, keyword, identifier string) error {
+	identity, err := i.idProvider.GetIdentity(token)
+	if err != nil {
+		return ErrNotAuthenticated
+	}
+	if !identity.Scopes.Contains(id.ScopeIndex) {
+		return ErrNotAuthorized
+	}
+
+	tagger, cryptor, err := i.getTaggerAndCryptor(keyword)
+	if err != nil {
+		return err
+	}
+
+	// Compute the current last sealed Identifier with the given keyword, i.e. the sealed
+	// Identifier with the largest value of the counter.
+	lastIdentifier, err := i.lastIdentifier(keyword)
+	if err != nil {
+		return err
+	}
+
+	counter := uint64(0)
+
+	// If the current last sealed Identifier is not an empty sealed Identifier, i.e. the keyword
+	// has already been added to the secure index, update the counter.
+	if lastIdentifier.Identifier != "" {
+		counter = lastIdentifier.NextCounter
+	}
+
+	// Compute new labelUUID and plaintext Identifier, seal it, and send it to the IO Provider.
+	labelUUID, err := computeLabelUUID(counter, tagger)
+	if err != nil {
+		return err
+	}
+
+	plainID := data.Identifier{Identifier: identifier, NextCounter: counter + 1}
+
+	sealedID, err := plainID.Seal(labelUUID, cryptor)
+	if err != nil {
+		return err
+	}
+	if err = i.putSealedIdentifier(labelUUID, &sealedID); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Given a keyword, Search returns all decrypted Identifiers that the keyword is contained in.
+func (i *SecureIndex) Search(token, keyword string) ([]string, error) {
+	identity, err := i.idProvider.GetIdentity(token)
+	if err != nil {
+		return nil, ErrNotAuthenticated
+	}
+	if !identity.Scopes.Contains(id.ScopeIndex) {
+		return nil, ErrNotAuthorized
+	}
+
+	tagger, cryptor, err := i.getTaggerAndCryptor(keyword)
+	if err != nil {
+		return nil, err
+	}
+
+	decryptedIDs := []string{}
+
+	// Starting with counter = 0, check if the corresponding keyword label exists in the secure index.
+	// As long as the keyword label exists, decrypt the corresponding Identifier, append it to decryptedIDs,
+	// and update counter to the next counter.
+	counter := uint64(0)
+
+	for {
+		labelUUID, err := computeLabelUUID(counter, tagger)
```

Comment:
For the sake of a smaller update later, I think we should compute the `tag` here and then call `computeLabelUUID` in the `put/get/deleteSealedIdentifier` functions. That way we only have to remove the conversion there when we update the IO provider.